graphistry/scrape_utils.py

```diff
@@ -0,0 +1,357 @@
+		else:
+			logger.info(f"{url_or_document_or_resource_or_pdf} not found in Tracker")
+			return EMPTY_DATAFRAME
+
+	def _url_to_resource(self, url):
+		resource = get_article_information(url)
+		if resource:
+			self.url = url
+			self.hashes.append(self.hash(url))
+			self.url_counts.update(
+				[url]
+			)  # how many times do you visit the url (if you care about changing resources, for example)
+			self._extract_resource(resource)
+
+	def _document_to_resource(self, document):
+		# takes care if the data is just plain text that should be parsed downstream
+		self.url = None
+		self.hashes.append(self.hash(document))
+		doc_urls = extract_all_urls(document)
+		resource = BASE_SCHEMA.copy()
+		resource["document"] = document
+		resource["urls_from_document"] = doc_urls
+		resource["hashes"] = hash_document(document)
+		logger.info("Resource is a document")
+		return resource
+
+	def _pdf_to_resource(self, url_to_pdf):
+		resource = get_pdf_from_url(url_to_pdf)
+		if resource:
+			self.url = url_to_pdf
+			self.hashes.append(self.hash(url_to_pdf))
+			self.url_counts.update(
+				[url_to_pdf]
+			)  # how many times do you visit the url (if you care about changing resources, for example)
+			logger.info("Scraped PDF")
+			return resource
+
+	def _extract_resource(self, resource):
+		if resource is None:
+			logger.info("Resource is None")
+			return
+		if self.is_pdf(resource):
+			resource = self._pdf_to_resource(resource)
+		elif not self.is_resource(resource):
+			resource = self._document_to_resource(resource)
+		self.res = self._to_dataframe(resource)  # convert to dataframe
+		self._add_to_store()
+		self._add_session_metadata()
+
+	def _add_to_store(self):
+		if self.tracker_file:
+			with open(self.tracker_file, "a") as f:
+				if f.tell() == 0:
+					header = BASE_SCHEMA.keys()
+				else:
+					header = False
+				self.res.to_csv(f, header=header, index=False)
+			if self.url is not None:
+				logger.info(f"Added [ {self.url} ] to `{self.tracker_file}`")
+			else:
+				logger.info(f"Added [[ {self.res['document'].values[0][:60]}..]] to `{self.tracker_file}`")
+		else:
+			logger.info("No tracker file specified")
+
+	def _add_session_metadata(self):
+		# get previous session metadata if any, and add to it
+		metadata = self._get_session_metadata()
+		if isinstance(metadata['url_counts'], dict): # since it is a dict, you need to convert it to a Counter
+			metadata['url_counts'] = Counter(metadata['url_counts'])
+		metadata['url_counts'].update(self.url_counts)
+		metadata['hashes'].extend(self.hashes)
+		with open(self.tracker_file_metadata, "w") as f:
+			json.dump(metadata, f)
+	
+	def _get_session_metadata(self):
+		if os.path.exists(self.tracker_file_metadata):
+			with open(self.tracker_file_metadata, "r") as f:
+				metadata = json.load(f)
+			return metadata
+		return {'hashes':[], 'url_counts': Counter()}
+
+
+def ParallelScraper(urls, tracker_file, n_jobs=-1, *args, **kwargs):
```

Comment:
ok for now, though for server support, we'd want to switch to dask

Reply:
Also, may be good to separate processes vs threads here, as i'm guessing mix of compute threads vs async network io here, like `n` jobs with `k` threads. Can defer till dask impl. Relevant: async io interface: https://stackoverflow.com/questions/68360767/newspaper3k-get-articles-from-html-instead-of-url